crates/matrix-sdk-crypto/src/olm/account.rs

```diff
@@ -1128,32 +1122,77 @@ impl Account {
         Ok(())
     }
 
-    /// Try to decrypt an Olm message.
+    /// Try to decrypt a normal Olm message.
     ///
-    /// This try to decrypt an Olm message using all the sessions we share
+    /// This tries to decrypt an Olm message using all the sessions we share
     /// with the given sender.
-    async fn decrypt_with_existing_sessions(
+    async fn decrypt_normal_olm_message(
         store: &Store,
+        sender: &UserId,
         sender_key: Curve25519PublicKey,
         message: &OlmMessage,
-    ) -> OlmResult<Option<(Session, String)>> {
-        let s = store.get_sessions(&sender_key.to_base64()).await?;
+    ) -> OlmResult<(SessionType, String)> {
+        let OlmMessage::Normal(_) = message else {
+            panic!("decrypt_normal_olm_message called with a prekey message");
+        };
+
+        let existing_sessions = store.get_sessions(&sender_key.to_base64()).await?;
+        if let Some(ref sessions) = existing_sessions {
+            let sessions: &mut Vec<Session> = &mut *sessions.lock().await;
+            // Try to decrypt the message using each Session we share with the
+            // given curve25519 sender key.
+            for session in sessions {
+                if let Ok(p) = session.decrypt(message).await {
+                    // success!
+                    return Ok((SessionType::Existing(session.clone()), p));
+                } else {
+                    // An error here is completely normal, after all we don't know
+                    // which session was used to encrypt a message. We will log a
+                    // warning if no session was able to decrypt the message.
+                    continue;
+                }
+            }
+        }
+
+        let session_ids = if let Some(sessions) = existing_sessions {
```

Comment:
Any reason this is a separate `if let` instead of part of the above `if let`?

Reply:
just that it's a bit awkward to return the session_ids from the bigger `if let`. But combining them does make it nicer to iterate the session list. I'll do so.